[19.0][MIG] account_invoice_triple_discount: Migration to 19.0

lef-adhoc opened the pull request.

chienandalu committed:
account_invoice_triple_discount: New module

pedrobaeza committed:
account_invoice_triple_discount: Some improvements

xavierjimenez committed:
account_invoice_triple_discount

rschnapka committed:
Translated using Weblate (German)

Currently translated at 100,0% (4 of 4 strings)

Translation: account-invoicing-11.0/account-invoicing-11.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-11-0/account-invoicing-11-0-account_invoice_triple_discount/de/

Nikul-Chaudhary committed:
[MIG] account_invoice_triple_discount v11 to v12

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-account_invoice_triple_discount/pt_BR/

benwillig committed:
[MIG] account_invoice_triple_discount to v13

ferran-S73 committed:
[IMP] account_invoice_triple_discount: black, isort, prettier

dreispt committed:
[FIX] account_invoice_triple_discount: Developer Mode was required to use discounts

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_invoice_triple_discount/it/

ramiadavid committed:
[MIG] account_invoice_triple_discount: Migration to 15.0

jabelchi committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-15.0/account-invoicing-15.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-15-0/account-invoicing-15-0-account_invoice_triple_discount/ca/

ramiadavid committed:
[FIX] account_invoice_triple_discount: fix precision when compute tax with discount

(cherry picked from commit 5afdb213c8f87be17bb21401a45cbf05a3548aa9)

ramiadavid committed:
[MIG] account_invoice_triple_discount: Migration to 16.0

ramiadavid committed:
[FIX][16.0] account_invoice_triple_discount : Error when print reports

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/hr/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/es/

SirAionTech committed:
[FIX] account_invoice_triple_discount: Discount rounding

grindtildeath committed:
[REF] account_invoice_triple_discount: Consolidate discount in std field

Until now, the triple discount feature did use the standard Odoo field
as the first discount field, adding only extra fields for the second
and the third discount. This implied we had to override any function
using discount to consider the other discounts properly.

By adding an extra field discount1 to store the first discount, it allows
to redefine the standard discount field to a computed field that will
consolidate the triple discount from the other fields, and avoid the
need to redefine anything relying on the discount field as it will
already consider the triple discounts.

lmignon committed:
[FIX] account_invoice_triple_discount: Fix tax recompute on closed fiscal year

When computing the taxes and the totals on account.move.line it's required to prevent calls to the write method on account.move or triggering recompute of unexpected fields while we assign a temporary value to the discount field to use the aggregated one bedore calling the base implementation. Otherwise the update of the discount field will invalidate totals and ... on related accoun.move and trigger calls to the write method of account.move. Any call to the write method will trigger a set of check to see if the balance is still right or check if you try to write on a record prior to the fiscal year close date or .... At the end, every time you try to access to the taxes on the records, some errors could be unexpeced raised due to the trick a temporarily changing the discount value of the field while computing others fields.

In the future we should uses dedicated fields for 3 discount fields and make the native discount field a computed one with readonly=False. This approach would simplify a lot the code. See https://github.com/OCA/account-invoicing/pull/1638

lmignon committed:
[IMP] account_invoice_triple_discount: Improves tests

Uses CommonCase to take benefit of proper env initialization and avoid useless tacking records creation

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/it/

legalsylvain committed:
[IMP] account_invoice_triple_discount: Simplify account.move.line algorithm

[FIX] account_invoice_triple_discount: populate discount1 field to avoid inconsistency if database contains account.move.line with not null discount before the installation of the module

[REF] account_invoice_triple_discount: Add GRAP as coauthor of the module

[FIX] account_invoice_triple_discount: Manage Correct creation of account move lines, where 'discount' field is set

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/sv/

ramiadavid committed:
Add test

grindtildeath committed:
Write discount into discount 1 if not defined

In case this module is installed and sale module is also installed
but not sale_triple_discount, or in case sale_triple_discount is not
updated to its latest version, make sure that setting discount, without
having discount1 defined, will not mess with the triple discounts.

legalsylvain committed:
[FIX] account_invoice_fixed_discount, account_invoice_triple_discount:
- set modules as incompatible in the manifest. (For the time being, there is only a informative message in the README of the module account_invoice_fixed_discount)
- Run separated tests for both modules.

legalsylvain committed:
[FIX] account_invoice_triple_discount: remove digits on overloaded field 'discount' as the field is now a technical field that should not have limitation. For exemple in the default case where digits precision of Discount is 2, if discounts are 05%, 09% and 13% the main discount is 24.7885 % (and not 24.79)

legalsylvain committed:
[REF] account_invoice_triple_discount: Improve speed computation of the field 'discount' and reduce complexity

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (11 of 11 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/it/

Borruso committed:
[IMP] account_invoice_triple_discount: pre-commit auto fixes

Borruso committed:
[MIG] account_invoice_triple_discount: Migration to 18.0

legalsylvain committed:
[REF] account_invoice_triple_discount : integrate triple.discount.mixin, previously in purchase_triple_discount module. (OCA/purchase-workflow)

legalsylvain committed:
[REF] account.move.line now inherit triple.discount.mixin model

legalsylvain committed:
[REF] account_invoice_triple_discount: single line writing

Commit:
[IMP]account_invoice_triple_discount:remove groups from tests

Commit:
[UPD] Update account_invoice_triple_discount.pot

OCA-git-bot committed:
[BOT] post-merge updates

tsezgin committed:
Translated using Weblate (Turkish)

Currently translated at 100.0% (11 of 11 strings)

Translation: account-invoicing-18.0/account-invoicing-18.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-18-0/account-invoicing-18-0-account_invoice_triple_discount/tr/

lef-adhoc committed:
[MIG] account_invoice_triple_discount: Migration to 19.0

luisDIXMIT reviewed (commented):
Should discounts be printed on the report? I’ve set the three discounts, but none of them are displayed on the report.

<img width="940" height="193" alt="image" src="https://github.com/user-attachments/assets/5c7cd9e3-03f7-4b16-adc9-70779381c8a0" />

ferran-S73 commented on account_invoice_triple_discount/report/invoice.xml:
```diff
@@ -0,0 +1,58 @@
+    >
+        <xpath expr="//t[@t-set='display_discount']" position="attributes">
+            <attribute name="t-value">False</attribute>
+        </xpath>
```
Not sure if this is something specific of this version but since there's four discount fields shouldn't this be as it is in account module so that discounts are shown if `discount` field isn't empty (which means that some of discount1, discount2 o discount3 aren't empty)

lef-adhoc commented:
Hi @luisDIXMIT @ferran-S73 

Fixed! The issue was that `display_discount` was set to `False`. I've updated it to check if any line has discount values (discount1, discount2, or discount3), similar to how the standard account module handles it. The discounts should now appear in the report.

ferran-S73 reviewed (approved):
code review

luisDIXMIT reviewed (approved):
Functionally tested and LGTM!

ferran-S73 commented on account_invoice_triple_discount/report/invoice.xml:
```diff
@@ -0,0 +1,61 @@
+                t-value="'text-end %s' % ('d-none d-md-table-cell' if report_type == 'html' else '')"
+            />
+        </xpath>
+        <xpath expr="//th[@name='th_discount']" position="attributes">
+            <attribute name="style">visibility: hidden;</attribute>
+        </xpath>
```
I was testing this functionally and found out that the `visibility: hidden` style is printing a gap inside the report's columns. I think this can be avoided by using a `t-if` condition instead. Please check this before merging as I cannot dismiss my own review

etobella commented:
@lef-adhoc there are some comments from @ferran-S73  , can you review?

lef-adhoc commented:
@ferran-S73 @etobella Done!

etobella reviewed (approved):
/ocabot merge nobump

lef-adhoc commented:
@etobella Do you know what the mistake is?

etobella commented:
there is incompatibilites with other modules. You might need to use copier update to split the workflow

ferran-S73 commented on account_invoice_triple_discount/report/invoice.xml:
```diff
@@ -0,0 +1,61 @@
+            <attribute name="t-if">False</attribute>
+        </xpath>
+        <xpath expr="//th[@name='th_discount']" position="after">
+            <th
+                name="th_discount1"
+                t-if="display_discount"
+                t-att-class="discount_class"
+            >
+                <span>Disc.1 %</span>
+            </th>
+            <th
+                name="th_discount2"
+                t-if="display_discount"
+                t-att-class="discount_class"
+            >
+                <span>Disc.2 %</span>
+            </th>
+            <th
+                name="th_discount3"
+                t-if="display_discount"
+                t-att-class="discount_class"
+            >
+                <span>Disc.3 %</span>
+            </th>
```
@lef-adhoc Adding two columns is breaking section lines appearance. Odoo computes a variable `line_colspan` in order to know hoy many cols to reserve to the section name based on how many columns are shown in the report. https://github.com/odoo/odoo/blob/5aa6711fcdcd28f09856eda442d510393ec46526/addons/account/views/report_invoice.xml#L223
This is helpful as it ensures the subtotal column value is printed under it's header.

Could you please add an xpath modifying the value of `line_colspan`. 2 should be added to it in case `display_discount` is True

eugenios73 commented:
@etobella @dreispt @luisDIXMIT @ferran-S73 superseded by https://github.com/OCA/account-invoicing/pull/2275 solving last comments

lef-adhoc commented:
@etobella @ferran-S73 Is that okay? Is it ready?

etobella reviewed (approved):
LGTM

luisDIXMIT reviewed (approved):
LGTM!

eugenios73 reviewed (approved):
LGFM

@legalsylvain @pedrobaeza Can someone please merge this PR? Thanks

lef-adhoc commented:
@etobella Can we merge it now?

etobella commented:
/ocabot merge nobump
